src/abstraction/transitory.rs

```diff
@@ -0,0 +1,218 @@
+    pub context: Context,
+    root_key_handle: ESYS_TR,
+}
+
+impl TransientObjectContext {
+    pub fn new(
+        tcti: Tcti,
+        root_key_size: usize,
+        root_key_auth_size: usize,
+        owner_hierarchy_auth: &[u8],
+    ) -> Result<Self> {
+        if root_key_auth_size > 32 {
+            return Err(Tss2ResponseCode::new(TPM2_RC_SIZE));
+        }
+        if root_key_size < 1024 {
+            return Err(Tss2ResponseCode::new(TPM2_RC_KEY_SIZE));
+        }
+        let mut context = Context::new(tcti)?;
+        let root_key_auth: Vec<u8> = if root_key_auth_size > 0 {
+            context.get_random(NO_SESSIONS, root_key_auth_size)?
+        } else {
+            vec![]
+        };
+        if !owner_hierarchy_auth.is_empty() {
+            context.set_handle_auth(ESYS_TR_RH_OWNER, owner_hierarchy_auth)?;
+        }
+
+        let root_key_handle = context.create_primary_key(
+            NO_SESSIONS,
+            ESYS_TR_RH_OWNER,
+            &get_rsa_public(true, true, false, root_key_size.try_into().unwrap()),
+            &root_key_auth,
+            &[],
+            &[],
+            &[],
+        )?;
+
+        let new_session = context.start_auth_session(
+            NO_NON_AUTH_SESSIONS,
+            root_key_handle,
+            ESYS_TR_NONE,
+            &[],
+            TPM2_SE_HMAC,
+            utils::TpmtSymDefBuilder::aes_256_cfb(),
+            TPM2_ALG_SHA256,
+        )?;
+        let session_attr = utils::TpmaSession::new()
+            .with_flag(TPMA_SESSION_DECRYPT)
+            .with_flag(TPMA_SESSION_ENCRYPT);
+        context.set_session_attr(new_session, session_attr)?;
+        context.set_session(new_session);
+        Ok(TransientObjectContext {
+            context,
+            root_key_handle,
+        })
+    }
+
+    pub fn create_rsa_signing_key(
+        &mut self,
+        key_size: usize,
+        auth_size: usize,
+    ) -> Result<(TpmsContext, AuthValue)> {
+        if auth_size > 32 {
+            return Err(Tss2ResponseCode::new(TPM2_RC_SIZE));
+        }
+        if key_size < 1024 {
+            return Err(Tss2ResponseCode::new(TPM2_RC_KEY_SIZE));
+        }
+        let key_auth = if auth_size > 0 {
+            self.context.get_random(NO_SESSIONS, 32)?
+        } else {
+            vec![]
+        };
+        let (key_priv, key_pub) = self.context.create_key(
+            NO_SESSIONS,
+            self.root_key_handle,
+            &get_rsa_public(false, false, true, key_size.try_into().unwrap()),
+            &key_auth,
+            &[],
+            &[],
+            &[],
+        )?;
+        let key_handle = self
+            .context
+            .load(NO_SESSIONS, self.root_key_handle, key_priv, key_pub)?;
+
+        let key_context = self.context.context_save(key_handle)?;
+        self.context.flush_context(key_handle)?;
+        Ok((key_context, key_auth))
+    }
+
+    pub fn load_external_rsa_public_key(&mut self, public_key: &[u8]) -> Result<TpmsContext> {
```

Comment:
Is there a way to set an `auth_value` for an imported public key? Or is it fine not to have one?

Reply:
Unfortunately, the only way to set an auth value is to provide a sensitive part to the object, which means it can no longer be loaded into the Owner hierarchy - only the NULL hierarchy allows uploading objects with both public and private sections. 

Since we want to keep the keys persistent, we have to stick to the OH, but that's probably ok - we have mutual authentication for the commands on the key, it's just that someone who gets to know the key handle in the short timeframe when it's in the TPM can use it to verify signatures.

Reply:
Ok! Thanks for the explanation.